### PR TITLE
GazeboDrone project added to connect a gazebo drone to the AirSim drone

### LINKED ISSUE #2380: Using `gzserver` (Gazebo backend) with Airsim
**What I'm trying to do**: I'm trying to use Airsim only as a frontend visualization tool and use Gazebo backend for modeling and physics. (reason being I've a lot of custom plugins developed in Gazebo at the moment) 

**Question**: Is there a way to achieve this in Airsim? If yes, how would one go about interfacing gazebo data with Airsim?

### PR DESCRIPTION
**Pull request by jonyMarino**
Fixes: #2380    

## About

This CMake project allows connecting a gazebo drone to the AirSim drone, using the gazebo drone as a flight dynamic model (FDM) and AirSim to generate sensor data. 

## How Has This Been Tested?

Following the added Readme.md
and running PX4 with `make px4_sitl gazebo_iris` together with QGroundControl.
Airsim using the compiled Blocks binary.

## Screenshots (if appropriate):

https://user-images.githubusercontent.com/4614600/120532801-7471a080-c3b6-11eb-8ccc-f98a910e5106.mp4

### CODE REVIEW

File: GazeboDrone/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+#include <iostream>
+
+constexpr int NWIDTH = 7;
+static constexpr int MESSAGE_THROTTLE = 100;
+
+using namespace msr::airlib;
+
+msr::airlib::MultirotorRpcLibClient client;
+
+void cbLocalPose(ConstPosesStampedPtr& _msg)
```
Comment (zimmy87) [inline GazeboDrone/src/main.cpp:44]: any reason for the underscore prefix here? same for `cbGobalPose()`

File: GazeboDrone/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    // Load gazebo
+    gazebo::client::setup(_argc, _argv);
+
+    // Create our node for communication
+    gazebo::transport::NodePtr node(new gazebo::transport::Node());
+    node->Init();
+
+    // Listen to Gazebo topics
+    gazebo::transport::SubscriberPtr subPose1 = node->Subscribe("~/pose/local/info", cbLocalPose);
+    gazebo::transport::SubscriberPtr subPose2 = node->Subscribe("~/pose/info", cbGobalPose);
```
Comment (zimmy87) [inline GazeboDrone/src/main.cpp:131]: locals should have under_scored style

File: GazeboDrone/src/main.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        if (i == _msg->pose_size() - 1) {
+            msr::airlib::Vector3r p;
+            msr::airlib::Quaternionr o;
+            p.x() = x;
+            p.y() = -y;
+            p.z() = -z;
+            o.w() = ow;
+            o.x() = ox;
+            o.y() = -oy;
+            o.z() = -oz;
```
Comment (zimmy87) [inline GazeboDrone/src/main.cpp:79]: You can simplify this by calling constructors for p & o

**Review by zimmy87 — CHANGES_REQUESTED** [review]
looks good other than some minor comments

File: GazeboDrone/src/main.cpp (hunk 1 above)
Comment (jonyMarino) [reply GazeboDrone/src/main.cpp]: ready

File: GazeboDrone/src/main.cpp (hunk 2 above)
Comment (jonyMarino) [reply GazeboDrone/src/main.cpp]: ready

File: GazeboDrone/src/main.cpp (hunk 3 above)
Comment (jonyMarino) [reply GazeboDrone/src/main.cpp]: ready

**Review by zimmy87 — APPROVED** [review]

**Comment by nkbeebe** [comment]
Is it feasible to replace the gazebo drone for a fixed wing model, to run the physics behind a plane, but still use AirSim for the data collection?

**Comment by jonyMarino** [comment]
@nkbeebe. Yes, you can do that. It would be great if someone contributes with an API method to set the rotors' angular velocity. Take a look at this tutorial on how to change the drone mesh: https://youtu.be/Bp86WiLUC80. You need to get the mesh of the vehicle you are trying to simulate. You would help the community a lot if you post a video on youtube on how to run AirSim with FixedWing in the way you decide to do it. Keep me updated please.

**Comment by nkbeebe** [comment]
But is that using Gazebo physics of a drone? Akin to simply replacing the Gazebo drone with a plane, but moving the plane as a drone would? My question is more of using the physics engine of a Gazebo plane, alongside the mesh. So the plane moves in the dynamics a Gazebo plane would move - not a gazebo drone would.

**Comment by jonyMarino** [comment]
Yes, I have understood your question. I'm not 100% sure if it is going to work, I think so. The GazeboDrone project subscribes to the pose updates of the drone and sends them to the AirSim drone through setVehiclePose, so if you use a plane in gazebo and the GazeboDrone project, can subscribe to its pose, then it will work.

**Comment by nikitabeebe** [comment]
Sorry if this is a simple question, but following the readme for this merge, I need to have UE4 running with the original AirSim build using clang, then once that is up along with gazebo, I can use the command ./GazeboDrone to then start that simulation? And for the build step, what directory am I creating a new build directory? (In the modified airsim file, replacing the original build folder?)

**Comment by jonyMarino** [comment]
Hi @nikitabeebe! You are correct about the order.
The build folder goes inside the GazeboDrone folder.

**Comment by nikitabeebe** [comment]
Also, does this project require the use of PX4, or can it be run without that program entirely? (Just using QGroundControl for joystick input)

**Comment by jonyMarino** [comment]
Hi @nikitabeebe! I think QGroundControl only works with PX4. But yes, you should be able to connect to gazebo without PX4 connected to it.

**Comment by nikitabeebe** [comment]
Thank you for confirming. That's what I initially thought @jonyMarino. If you could then, can you take a look at an issue request I created, as I am still unable to compile and run this project like you have demonstrated. I've tried to rule out all the issues with versions or missing dependencies, but I keep getting hung up on the make task for building the GazeboDrone...

https://github.com/microsoft/AirSim/issues/3874

Any input or insight into possible fixes would be a life saver!

**Pull request merged**